### ISSUE TITLE
Document AdjointLayout container

### ISSUE THREAD

**Issue by philippjfr** [ioam/holoviews#1129]
As was noticed in https://github.com/ioam/holoviews/issues/358, the Containers tutorial does not currently explain AdjointLayouts.

**Comment by jlstevens** [comment]
It seems we are stuck with ``AdjointLayouts`` because they are useful in practice, even though they aren't particularly elegant (conceptually or in terms of implementation).

Do we have any plans for how they could be improved eventually?

**Comment by philippjfr** [comment]
Right, the concept of adjoining plots is **very** useful as it has turned out to visualize marginal distributions or other statistics in various ways. The current implementation is indeed inelegant as you can only adjoin two plots, and if only one is adjoined it must be on the right. Making that more general should be an eventual goal but it shouldn't stop us documenting the current behavior.

**Comment by jlstevens** [comment]
Here are the issues I have with AdjointLayout (off the top of my head):

* If is a special type of Layout that nests into normal Layouts which makes it somewhat element-like and means it doesn't fit naturally into the hierarchy...
* ... but you can only really nest it once. Even if the implementation allowed it, multiple nestings would quickly become illegible.
* It has its own, poorly documented operator ``<<`` which I would like to deprecate. But what would it be replaced with?
* As you mentioned, there is ad hoc behavior where it joins on the right, then on the top? then on the left? I can't remember, which shows how arbitrary it is!
* It complicates the layout code which is plenty complicated already.

I don't mind documenting ``AdjointLayout`` as long as I know how we intend to transition to a better system in future. The reason it is poorly documented is because we don't like it as devs and feel there should be a better system!

What I would like to see is a flexible layout system/specification that covers insets and ``AdjointLayout`` without resulting in a horrendously complicated. Ideally, we would have a layout solver similar to the bokeh layout solver (without depending on bokeh), perhaps something like [this](https://github.com/slightlyoff/cassowary.js/).

**Comment by dhruvbalwada** [comment]
Just wondering if there is any progress on this issue? 

The behavior of the AdjointLayouts that are automatically created is completely unintuitive at the moment, and up to trial and error. I opened a discourse issue about this: https://discourse.holoviz.org/t/inconsistent-behavior-of-colormap-for-adjoint-layout-with-hist/1925 

In principle adding a histogram to a plot, or other adjoint plots, seems like a very powerful tool. You could add some documentation about it, and just say that in future version it will be incorporated in a different way. 

This lack of documentation is causing some of us to spend a lot of time trying to reverse engineer things and apply mods to try and get the behavior we want. Here is a half working example, trying to get things to work with rasterize to look like they would with holomaps:
```
import holoviews as hv
from holoviews import opts
from holoviews.streams import Stream
import param
import xarray as xr
import hvplot.xarray
import numpy as np
import panel as pn
hv.extension('bokeh')
print(hv.__version__)
print(pn.__version__)
print(hvplot.__version__)
#print(xr.show_versions())
print(xr.__version__)
# Data
ls = np.linspace(0, 10, 200)
xx, yy = np.meshgrid(ls, ls)
z = np.sin(xx)*np.cos(yy)
da = xr.DataArray(z, coords=[ls, ls], dims=["y", "x"], name='test')
da


# Visualization
class Style(param.Parameterized):
    colormap = param.ObjectSelector(default='viridis', objects=['viridis', 'plasma', 'magma'])
    color_levels = param.Integer(default=255, bounds=(1, 255))
    # Does not work to seperate out, need to link somehow
    #image = da.hvplot.image(rasterize=True)
    #hist = da.hvplot.hist() 
    # Some weirdness in that reset button does not return to first axes extent after panning and zooming
    image, hist = da.hvplot.image(rasterize=True).hist() #AdjointLayout does some linking magic between image and hist
    rangexy = hv.streams.RangeXY(source=image)
    rangez = hv.streams.BoundsY(boundsy=(-1,1), source=hist) # doesn't work well when resetting histogram plot
    #rangez = hv.streams.RangeX(x_range=(-1,1), source=hist) #needs a default value or else None
    
    def _set_tools(self, plot, element):
        plot.state.toolbar.active_drag = None
        plot.state.toolbar.active_inspect = None
        
    def histogram_view(self):
        myhist = self.hist.opts(color='gray', tools=['ybox_select'], invert_axes=True, width=150
                                #active_tools=['wheel_zoom'],
                             )
        return myhist
    
    @param.depends('colormap','color_levels', 'rangez.param')
    def image_view(self):
        myimage = self.image.opts(colorbar=True, 
                          width=400, 
                          clim=self.rangez.boundsy,
                          cmap=self.colormap, 
                          color_levels=self.color_levels, 
                          tools=['hover'], 
                          toolbar='above',
                          hooks=[self._set_tools])
        return myimage
    def panel(self):
        #wrap functions that return holoviews things in pn.Something() to call myplot.panel() or myplot.view()
        # how do we get colorbar to work with histogram
        composition = pn.Row(self.image_view) + pn.Row(self.histogram_view) 
        return composition
myplot = Style()
myplot.panel()
```
